% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{BaseSpacePy Documentation}
\date{March 12, 2014}
\release{0.2}
\author{Illumina}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\def\PYG@tok@gd{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\def\PYG@tok@gu{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\def\PYG@tok@gt{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.25,0.82}{##1}}}
\def\PYG@tok@gs{\let\PYG@bf=\textbf}
\def\PYG@tok@gr{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\def\PYG@tok@cm{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\def\PYG@tok@vg{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\def\PYG@tok@m{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@mh{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@cs{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\colorbox[rgb]{1.00,0.94,0.94}{##1}}}
\def\PYG@tok@ge{\let\PYG@it=\textit}
\def\PYG@tok@vc{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\def\PYG@tok@il{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@go{\def\PYG@tc##1{\textcolor[rgb]{0.19,0.19,0.19}{##1}}}
\def\PYG@tok@cp{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@gi{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\def\PYG@tok@gh{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\def\PYG@tok@ni{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\def\PYG@tok@nl{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\def\PYG@tok@nn{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\def\PYG@tok@no{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\def\PYG@tok@na{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@nb{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@nc{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\def\PYG@tok@nd{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\def\PYG@tok@ne{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@nf{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\def\PYG@tok@si{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\def\PYG@tok@s2{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@vi{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\def\PYG@tok@nt{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\def\PYG@tok@nv{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\def\PYG@tok@s1{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@gp{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\def\PYG@tok@sh{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@ow{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@sx{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\def\PYG@tok@bp{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@c1{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\def\PYG@tok@kc{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@c{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\def\PYG@tok@mf{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@err{\def\PYG@bc##1{\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{##1}}}
\def\PYG@tok@kd{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@ss{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\def\PYG@tok@sr{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\def\PYG@tok@mo{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@mi{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\def\PYG@tok@kn{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@o{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\def\PYG@tok@kr{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@s{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@kp{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@w{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\def\PYG@tok@kt{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\def\PYG@tok@sc{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@sb{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@k{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\def\PYG@tok@se{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\def\PYG@tok@sd{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Contents:


\chapter{Available modules}
\label{Available modules:available-modules}\label{Available modules::doc}\label{Available modules:basespacepy}

\section{API}
\label{Available modules:api}\index{BaseSpaceAPI (class in BaseSpacePy.api.BaseSpaceAPI)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI}\pysiglinewithargsret{\strong{class }\code{BaseSpacePy.api.BaseSpaceAPI.}\bfcode{BaseSpaceAPI}}{\emph{clientKey=None}, \emph{clientSecret=None}, \emph{apiServer=None}, \emph{version=None}, \emph{appSessionId='`}, \emph{AccessToken='`}, \emph{timeout=10}, \emph{profile='DEFAULT'}}{}
The main API class used for all communication with the REST server
\index{appResultFileUpload() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.appResultFileUpload}\pysiglinewithargsret{\bfcode{appResultFileUpload}}{\emph{Id}, \emph{localPath}, \emph{fileName}, \emph{directory}, \emph{contentType}}{}
Uploads a file associated with an AppResult to BaseSpace and returns the corresponding file object.
Small files are uploaded with a single-part upload method, while larger files (\textless{} 25 MB) are uploaded
with multipart upload.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- AppResult id.

\item {} 
\textbf{localPath} -- The local path to the file to be uploaded, including file name.

\item {} 
\textbf{fileName} -- The desired filename in the AppResult folder on the BaseSpace server.

\item {} 
\textbf{directory} -- The directory the file should be placed in on the BaseSpace server.

\item {} 
\textbf{contentType} -- The content-type of the file, eg. `text/plain' for text files, `application/octet-stream' for binary files

\end{itemize}

\item[{Returns}] \leavevmode
a newly created File instance

\end{description}\end{quote}

\end{fulllineitems}

\index{createAppResult() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.createAppResult}\pysiglinewithargsret{\bfcode{createAppResult}}{\emph{Id}, \emph{name}, \emph{desc}, \emph{samples=None}, \emph{appSessionId=None}}{}
Create an AppResult object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the project in which the AppResult is to be added

\item {} 
\textbf{name} -- The name of the AppResult

\item {} 
\textbf{desc} -- A description of the AppResult

\item {} 
\textbf{samples} -- (Optional) A list of one or more Samples Ids that the AppResult is related to

\item {} 
\textbf{appSessionId} -- (Optional) If no appSessionId is given, the id used to initialize the BaseSpaceAPI instance will be used. If appSessionId is set equal to an empty string, a new appsession will be created for the appresult object

\end{itemize}

\item[{Raises Exception}] \leavevmode
when attempting to create AppResult in an AppSession that has a status other than `running'.

\item[{Returns}] \leavevmode
a newly created AppResult instance

\end{description}\end{quote}

\end{fulllineitems}

\index{createProject() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.createProject}\pysiglinewithargsret{\bfcode{createProject}}{\emph{Name}}{}
Creates a project with the specified name and returns a project object. 
If a project with this name already exists, the existing project is returned.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Name} -- Name of the project

\item[{Returns}] \leavevmode
a Project instance of the newly created project

\end{description}\end{quote}

\end{fulllineitems}

\index{fileDownload() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.fileDownload}\pysiglinewithargsret{\bfcode{fileDownload}}{\emph{Id}, \emph{localDir}, \emph{byteRange=None}, \emph{createBsDir=False}}{}
Downloads a BaseSpace file to a local directory, and names the file with the BaseSpace file name.
If the File has a directory in BaseSpace, it will be re-created locally in the provided localDir 
(to disable this, set createBsPath=False).

If the file is large, multi-part download will be used.

Byte-range requests are supported for only small byte ranges (single-part downloads).
Byte-range requests are restricted to a single request of `start' and `end' bytes, without support for
negative or empty values for `start' or `end'.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The file id

\item {} 
\textbf{localDir} -- The local directory to place the file in

\item {} 
\textbf{byteRange} -- (optional) The byte range of the file to retrieve, provide a 2-element list with start and end byte values

\item {} 
\textbf{createBsDir} -- (optional) create BaseSpace File's directory inside localDir (default: False)

\end{itemize}

\item[{Raises ByteRangeException}] \leavevmode
if the provided byte range is invalid

\item[{Returns}] \leavevmode
a File instance

\end{description}\end{quote}

\end{fulllineitems}

\index{fileS3metadata() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.fileS3metadata}\pysiglinewithargsret{\bfcode{fileS3metadata}}{\emph{Id}}{}
Returns the S3 url and etag (md5 for small files uploaded as a single part) for a BaseSpace file
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- The file id

\item[{Raises Exception}] \leavevmode
if REST API call to BaseSpace server fails

\item[{Returns}] \leavevmode
Dict with s3 url (`url' key) and etag (`etag' key)

\end{description}\end{quote}

\end{fulllineitems}

\index{fileUrl() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.fileUrl}\pysiglinewithargsret{\bfcode{fileUrl}}{\emph{Id}}{}
** Deprecated in favor of fileS3metadata() **

Returns URL of file (on S3)
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- The file id

\item[{Raises Exception}] \leavevmode
if REST API call to BaseSpace server fails

\item[{Returns}] \leavevmode
a URL

\end{description}\end{quote}

\end{fulllineitems}

\index{filterVariantSet() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.filterVariantSet}\pysiglinewithargsret{\bfcode{filterVariantSet}}{\emph{Id}, \emph{Chrom}, \emph{StartPos}, \emph{EndPos}, \emph{Format='json'}, \emph{queryPars=None}}{}
List the variants in a set of variants. Note the maximum returned records is 1000.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the variant file

\item {} 
\textbf{Chrom} -- Chromosome name

\item {} 
\textbf{StartPos} -- The start position of the sequence of interest

\item {} 
\textbf{EndPos} -- The start position of the sequence of interest

\item {} 
\textbf{Format} -- (optional) Format for results, possible values: `vcf' (not implemented yet), `json'(default, which actually returns an object)

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of Variant instances, when Format is json; a string, when Format is vcf

\end{description}\end{quote}

\end{fulllineitems}

\index{getAccess() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAccess}\pysiglinewithargsret{\bfcode{getAccess}}{\emph{obj}, \emph{accessType='write'}, \emph{web=False}, \emph{redirectURL='`}, \emph{state='`}}{}
Requests access to the provided BaseSpace object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{obj} -- The data object we wish to get access to -- must be a Project, Sample, AppResult, or Run.

\item {} 
\textbf{accessType} -- (Optional) the type of access (browse\textbar{}read\textbar{}write\textbar{}create), default is write. Create is only supported for Projects.

\item {} 
\textbf{web} -- (Optional) true if the App is web-based, default is false meaning a device based app

\item {} 
\textbf{redirectURL} -- (Optional) Redirect URL for the web-based case

\item {} 
\textbf{state} -- (Optional) A parameter that will passed through to the redirect response.

\end{itemize}

\item[{Raises ModelNotSupportedException}] \leavevmode
for classes of objects not supported by this method

\item[{Returns}] \leavevmode
for device requests, a dictionary of server response; for web requests, a url to to send the user to

\end{description}\end{quote}

\end{fulllineitems}

\index{getAccessibleRunsByUser() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAccessibleRunsByUser}\pysiglinewithargsret{\bfcode{getAccessibleRunsByUser}}{\emph{queryPars=None}}{}
Returns a list of accessible runs for the current User
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\item[{Returns}] \leavevmode
a list of Run instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResultById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppResultById}\pysiglinewithargsret{\bfcode{getAppResultById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns an AppResult object corresponding to Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the AppResult

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
an AppResult instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResultFiles() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppResultFiles}\pysiglinewithargsret{\bfcode{getAppResultFiles}}{\emph{Id}, \emph{queryPars=None}}{}~\begin{itemize}
\item {} 
Deprecated in favor of getAppResultFileById() *

\end{itemize}

Returns a list of File object for an AppResult
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the AppResult

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of File instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResultFilesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppResultFilesById}\pysiglinewithargsret{\bfcode{getAppResultFilesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns a list of File object for an AppResult
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the AppResult

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of File instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResultPropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppResultPropertiesById}\pysiglinewithargsret{\bfcode{getAppResultPropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns the Properties of an AppResult object corresponding to AppResult Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the AppResult

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a PropertyList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResultsByProject() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppResultsByProject}\pysiglinewithargsret{\bfcode{getAppResultsByProject}}{\emph{Id}, \emph{queryPars=None}, \emph{statuses=None}}{}
Returns a list of AppResult object associated with the project with Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The project id

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\item {} 
\textbf{statuses} -- An (optional) list of AppResult statuses to filter by, eg., `complete'

\end{itemize}

\item[{Returns}] \leavevmode
a list of AppResult instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppSession() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppSession}\pysiglinewithargsret{\bfcode{getAppSession}}{\emph{Id=None}}{}
Get metadata about an AppSession.         
Note that the client key and secret must match those of the AppSession's Application.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- an AppSession Id; if not provided, the AppSession Id of the BaseSpaceAPI instance will be used

\item[{Returns}] \leavevmode
An AppSession instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppSessionById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppSessionById}\pysiglinewithargsret{\bfcode{getAppSessionById}}{\emph{Id}}{}
Get metadata about an AppSession.
Note that the client key and secret must match those of the AppSession's Application.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- The Id of the AppSession

\item[{Returns}] \leavevmode
An AppSession instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppSessionInputsById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppSessionInputsById}\pysiglinewithargsret{\bfcode{getAppSessionInputsById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns the input properties of an AppSession
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- An AppSessionId

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a dictionary of input properties, keyed by input Name

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppSessionPropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppSessionPropertiesById}\pysiglinewithargsret{\bfcode{getAppSessionPropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns the Properties of an AppSession
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- An AppSession Id

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
A PropertyList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppSessionPropertyByName() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAppSessionPropertyByName}\pysiglinewithargsret{\bfcode{getAppSessionPropertyByName}}{\emph{Id}, \emph{name}, \emph{queryPars=None}}{}
Returns the multi-value Property of the provided AppSession that has the provided Property name.
Note - this method (and REST API) is supported for ONLY multi-value Properties.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The AppSessionId

\item {} 
\textbf{name} -- Name of the multi-value property to retrieve

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
A multi-value propertylist instance such as MultiValuePropertyAppResultsList (depending on the Property Type)

\end{description}\end{quote}

\end{fulllineitems}

\index{getAvailableGenomes() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getAvailableGenomes}\pysiglinewithargsret{\bfcode{getAvailableGenomes}}{\emph{queryPars=None}}{}
Returns a list of all available genomes
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\item[{Returns}] \leavevmode
a list of GenomeV1 instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getCoverageMetaInfo() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getCoverageMetaInfo}\pysiglinewithargsret{\bfcode{getCoverageMetaInfo}}{\emph{Id}, \emph{Chrom}}{}
Returns metadata about coverage of a chromosome.
Note that HrefCoverage must be available for the provided BAM file
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- the Id of a Bam file

\item {} 
\textbf{Chrom} -- chromosome name

\end{itemize}

\item[{Returns}] \leavevmode
a CoverageMetaData instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getFileById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getFileById}\pysiglinewithargsret{\bfcode{getFileById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns a file object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the file

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a File instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getFilePropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getFilePropertiesById}\pysiglinewithargsret{\bfcode{getFilePropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns the Properties of a file object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the file

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a PropertyList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getFilesBySample() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getFilesBySample}\pysiglinewithargsret{\bfcode{getFilesBySample}}{\emph{Id}, \emph{queryPars=None}}{}~\begin{itemize}
\item {} 
Deprecated in favor of getSampleFilesById() *

\end{itemize}

Returns a list of File objects associated with a Sample
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- A Sample id

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of File instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getGenomeById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getGenomeById}\pysiglinewithargsret{\bfcode{getGenomeById}}{\emph{Id}}{}
Returns an instance of Genome with the specified Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- The genome id

\item[{Returns}] \leavevmode
a GenomeV1 instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getIntervalCoverage() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getIntervalCoverage}\pysiglinewithargsret{\bfcode{getIntervalCoverage}}{\emph{Id}, \emph{Chrom}, \emph{StartPos}, \emph{EndPos}}{}
Returns metadata about an alignment, including max coverage and cov granularity.
Note that HrefCoverage must be available for the provided BAM file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- the Id of a BAM file

\item {} 
\textbf{Chrom} -- chromosome name

\item {} 
\textbf{StartPos} -- get coverage starting at this position

\item {} 
\textbf{EndPos} -- get coverage up to and including this position; the returned EndPos may be larger than requested due to rounding up to nearest window end coordinate

\end{itemize}

\item[{Returns}] \leavevmode
a Coverage instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getProjectById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getProjectById}\pysiglinewithargsret{\bfcode{getProjectById}}{\emph{Id}, \emph{queryPars=None}}{}
Request a project object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the project

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a Project instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getProjectByUser() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getProjectByUser}\pysiglinewithargsret{\bfcode{getProjectByUser}}{\emph{queryPars=None}}{}
Returns a list available projects for the current User.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\item[{Returns}] \leavevmode
a list of Project instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getProjectPropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getProjectPropertiesById}\pysiglinewithargsret{\bfcode{getProjectPropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Request the Properties of a project object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the project

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a ProjectList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getRunById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getRunById}\pysiglinewithargsret{\bfcode{getRunById}}{\emph{Id}, \emph{queryPars=None}}{}
Request a run object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the run

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a Run instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getRunFilesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getRunFilesById}\pysiglinewithargsret{\bfcode{getRunFilesById}}{\emph{Id}, \emph{queryPars=None}}{}
Request the files associated with a Run, using the Run's Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the run

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of Run instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getRunPropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getRunPropertiesById}\pysiglinewithargsret{\bfcode{getRunPropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Request the Properties of a run object by Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the run

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a PropertyList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getRunSamplesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getRunSamplesById}\pysiglinewithargsret{\bfcode{getRunSamplesById}}{\emph{Id}, \emph{queryPars=None}}{}
Request the Samples associated with a Run, using the Run's Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the run

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of Sample instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getSampleById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getSampleById}\pysiglinewithargsret{\bfcode{getSampleById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns a Sample object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the sample

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a Sample instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getSampleFilesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getSampleFilesById}\pysiglinewithargsret{\bfcode{getSampleFilesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns a list of File objects associated with a Sample
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- A Sample id

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of File instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getSamplePropertiesById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getSamplePropertiesById}\pysiglinewithargsret{\bfcode{getSamplePropertiesById}}{\emph{Id}, \emph{queryPars=None}}{}
Returns the Properties of a Sample object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the sample

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a PropertyList instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getSamplesByProject() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getSamplesByProject}\pysiglinewithargsret{\bfcode{getSamplesByProject}}{\emph{Id}, \emph{queryPars=None}}{}
Returns a list of samples associated with a project with Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the project

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of Sample instances

\end{description}\end{quote}

\end{fulllineitems}

\index{getUserById() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getUserById}\pysiglinewithargsret{\bfcode{getUserById}}{\emph{Id}}{}
Returns the User object corresponding to User Id
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{Id} -- The Id of the user

\item[{Returns}] \leavevmode
a User instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getVariantMetadata() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getVariantMetadata}\pysiglinewithargsret{\bfcode{getVariantMetadata}}{\emph{Id}, \emph{Format='json'}}{}
Returns the header information of a VCF file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The Id of the VCF file

\item {} 
\textbf{Format} -- (optional) The return-value format, set to `json' (default) to return return an object (not actually json format), or `vcf' (not implemented yet) to return a string in VCF format.

\end{itemize}

\item[{Returns}] \leavevmode
A VariantHeader instance

\end{description}\end{quote}

\end{fulllineitems}

\index{getVerificationCode() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getVerificationCode}\pysiglinewithargsret{\bfcode{getVerificationCode}}{\emph{scope}}{}
For non-web applications (eg. devices), returns the device code 
and verification url for the user to approve access to a specific data scope.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{scope} -- The scope that access is requested for (e.g. `browse project 123')

\item[{Returns}] \leavevmode
dictionary of server response

\end{description}\end{quote}

\end{fulllineitems}

\index{getWebVerificationCode() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.getWebVerificationCode}\pysiglinewithargsret{\bfcode{getWebVerificationCode}}{\emph{scope}, \emph{redirectURL}, \emph{state='`}}{}
Generates the URL the user should be redirected to for web-based authentication
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{scope} -- The scope that access is requested for (e.g. `browse project 123')

\item {} 
\textbf{redirectURL} -- The redirect URL

\item {} 
\textbf{state} -- (Optional) A state parameter that will passed through to the redirect response

\end{itemize}

\item[{Returns}] \leavevmode
a url

\end{description}\end{quote}

\end{fulllineitems}

\index{multipartFileDownload() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.multipartFileDownload}\pysiglinewithargsret{\bfcode{multipartFileDownload}}{\emph{Id}, \emph{localDir}, \emph{processCount=10}, \emph{partSize=25}, \emph{createBsDir=False}, \emph{tempDir='`}}{}
Method for multi-threaded file-download for parallel transfer of very large files (currently only runs on unix systems)
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The ID of the File to download

\item {} 
\textbf{localDir} -- The local path in which to store the downloaded file

\item {} 
\textbf{processCount} -- (optional) The number of processes to be used, default 10

\item {} 
\textbf{partSize} -- (optional) The size in MB of individual file parts to download, default 25

\item {} 
\textbf{createBsDir} -- (optional) create BaseSpace File's directory in local\_dir, default False

\item {} 
\textbf{tempDir} -- (optional) Set temp directory to use debug mode, which stores downloaded file chunks in individual files, then completes by `cat'ing chunks into large file

\end{itemize}

\item[{Returns}] \leavevmode
a File instance

\end{description}\end{quote}

\end{fulllineitems}

\index{multipartFileUpload() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.multipartFileUpload}\pysiglinewithargsret{\bfcode{multipartFileUpload}}{\emph{Id}, \emph{localPath}, \emph{fileName}, \emph{directory}, \emph{contentType}, \emph{tempDir=None}, \emph{processCount=10}, \emph{partSize=25}}{}
Method for multi-threaded file-upload for parallel transfer of very large files (currently only runs on unix systems)
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The AppResult ID

\item {} 
\textbf{localPath} -- The local path of the file to upload, including file name; local path will not be stored in BaseSpace (use directory argument for this)

\item {} 
\textbf{fileName} -- The desired filename on the server

\item {} 
\textbf{directory} -- The desired directory name on the server (empty string will place it in the root directory)

\item {} 
\textbf{contentType} -- The content type of the file

\item {} 
\textbf{tempdir} -- (optional) Temp directory to use for temporary file chunks to upload

\item {} 
\textbf{processCount} -- (optional) The number of processes to be used, default 10

\item {} 
\textbf{partSize} -- (optional) The size in MB of individual upload parts (must be \textgreater{}5 Mb and \textless{}=25 Mb), default 25

\end{itemize}

\item[{Returns}] \leavevmode
a File instance, which has been updated after the upload has completed.

\end{description}\end{quote}

\end{fulllineitems}

\index{obtainAccessToken() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.obtainAccessToken}\pysiglinewithargsret{\bfcode{obtainAccessToken}}{\emph{code}, \emph{grantType='device'}, \emph{redirect\_uri=None}}{}
Returns a user specific access token, for either device (non-web) or web apps.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{code} -- The device code returned by the getVerificationCode method

\item {} 
\textbf{grantType} -- Grant-type may be either `device' for non-web apps (default), or `authorization\_code' for web apps

\item {} 
\textbf{redirect\_uri} -- The uri to redirect to; required for web apps only.

\end{itemize}

\item[{Raises OAuthException}] \leavevmode
when redirect\_uri isn't provided by web apps

\item[{Returns}] \leavevmode
an access token

\end{description}\end{quote}

\end{fulllineitems}

\index{setAppSessionState() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.setAppSessionState}\pysiglinewithargsret{\bfcode{setAppSessionState}}{\emph{Id}, \emph{Status}, \emph{Summary}}{}
Set the Status and StatusSummary of an AppSession in BaseSpace.
Note - once Status is set to Completed or Aborted, no further changes can made.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{Id} -- The id of the AppSession

\item {} 
\textbf{Status} -- The AppSession status string, must be one of: Running, Complete, NeedsAttention, TimedOut, Aborted

\item {} 
\textbf{Summary} -- The status summary string

\end{itemize}

\item[{Returns}] \leavevmode
An updated AppSession instance

\end{description}\end{quote}

\end{fulllineitems}

\index{updatePrivileges() (BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.api.BaseSpaceAPI.BaseSpaceAPI.updatePrivileges}\pysiglinewithargsret{\bfcode{updatePrivileges}}{\emph{code}, \emph{grantType='device'}, \emph{redirect\_uri=None}}{}
Retrieves a user-specific access token, and sets the token on the current object.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{code} -- The device code returned by the getVerificationCode method

\item {} 
\textbf{grantType} -- Grant-type may be either `device' for non-web apps (default), or `authorization\_code' for web apps

\item {} 
\textbf{redirect\_uri} -- The uri to redirect to; required for web apps only.

\end{itemize}

\item[{Returns}] \leavevmode
None

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\section{Models}
\label{Available modules:models}

\subsection{Project}
\label{Available modules:project}\index{Project (class in BaseSpacePy.model.Project)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project}\pysigline{\strong{class }\code{BaseSpacePy.model.Project.}\bfcode{Project}}
Represents a BaseSpace Project object.
\index{createAppResult() (BaseSpacePy.model.Project.Project method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project.createAppResult}\pysiglinewithargsret{\bfcode{createAppResult}}{\emph{api}, \emph{name}, \emph{desc}, \emph{samples=None}, \emph{appSessionId=None}}{}
Return a newly created app result object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{name} -- The name of the app result

\item {} 
\textbf{desc} -- A description of the app result

\item {} 
\textbf{samples} -- (Optional) A list of one or more Samples Ids that the AppResult is related to

\item {} 
\textbf{appSessionId} -- (Optional) If no appSessionId is given, the id used to initialize the BaseSpaceAPI instance will be used. If appSessionId is set equal to an empty string, a new appsession will be created for the appresult object

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{getAccessStr() (BaseSpacePy.model.Project.Project method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project.getAccessStr}\pysiglinewithargsret{\bfcode{getAccessStr}}{\emph{scope='write'}}{}
Returns the scope-string to used for requesting BaseSpace access to the object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{scope} -- The scope-type that is request (write\textbar{}read)

\end{description}\end{quote}

\end{fulllineitems}

\index{getAppResults() (BaseSpacePy.model.Project.Project method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project.getAppResults}\pysiglinewithargsret{\bfcode{getAppResults}}{\emph{api}, \emph{queryPars=None}, \emph{statuses=None}}{}
Returns a list of AppResult objects.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\item {} 
\textbf{statuses} -- An optional list of statuses, eg. `complete'

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{getSamples() (BaseSpacePy.model.Project.Project method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project.getSamples}\pysiglinewithargsret{\bfcode{getSamples}}{\emph{api}, \emph{queryPars=None}}{}
Returns a list of Sample objects.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{isInit() (BaseSpacePy.model.Project.Project method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Project.Project.isInit}\pysiglinewithargsret{\bfcode{isInit}}{}{}
Tests if the Project instance has been initialized.
\begin{quote}\begin{description}
\item[{Raises ModelNotInitializedException}] \leavevmode
if the Id variable is not set

\item[{Returns}] \leavevmode
True on success

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subsection{AppSession and AppResult}
\label{Available modules:appsession-and-appresult}\index{AppSession (class in BaseSpacePy.model.AppSession)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppSession.AppSession}\pysigline{\strong{class }\code{BaseSpacePy.model.AppSession.}\bfcode{AppSession}}
Returned from getAppSessionById() and getAppSesssion()

\end{fulllineitems}

\index{AppResult (class in BaseSpacePy.model.AppResult)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult}\pysigline{\strong{class }\code{BaseSpacePy.model.AppResult.}\bfcode{AppResult}}~\index{getAccessStr() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.getAccessStr}\pysiglinewithargsret{\bfcode{getAccessStr}}{\emph{scope='write'}}{}
Returns the scope-string to be used for requesting BaseSpace access to the object
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{scope} -- The scope-type that is request (write\textbar{}read)

\end{description}\end{quote}

\end{fulllineitems}

\index{getFiles() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.getFiles}\pysiglinewithargsret{\bfcode{getFiles}}{\emph{api}, \emph{queryPars=None}}{}
Returns a list of file objects
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{getReferencedSamples() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.getReferencedSamples}\pysiglinewithargsret{\bfcode{getReferencedSamples}}{\emph{api}}{}
Returns the sample objects for the referenced sample(s). 
NOTE this method makes one request to REST server per sample.
If other reference types than Samples are present (they shouldn't be), they are ignored.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{api} -- A BaseSpaceAPI instance

\item[{Returns}] \leavevmode
A list of sample objects that are referenced by the AppResult.

\end{description}\end{quote}

\end{fulllineitems}

\index{getReferencedSamplesIds() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.getReferencedSamplesIds}\pysiglinewithargsret{\bfcode{getReferencedSamplesIds}}{}{}
Return the sample ids for the referenced sample(s). 
If other reference types than Samples are present (they shouldn't be), they are ignored.
\begin{quote}\begin{description}
\item[{Returns}] \leavevmode
a list of sample ids for the referenced samples.

\end{description}\end{quote}

\end{fulllineitems}

\index{isInit() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.isInit}\pysiglinewithargsret{\bfcode{isInit}}{}{}
Tests if the AppResult instance has been initialized.
\begin{quote}\begin{description}
\item[{Raises ModelNotInitializedException}] \leavevmode
if the Id variable is not set

\item[{Returns}] \leavevmode
True on success

\end{description}\end{quote}

\end{fulllineitems}

\index{uploadFile() (BaseSpacePy.model.AppResult.AppResult method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.AppResult.AppResult.uploadFile}\pysiglinewithargsret{\bfcode{uploadFile}}{\emph{api}, \emph{localPath}, \emph{fileName}, \emph{directory}, \emph{contentType}}{}
Uploads a local file to the BaseSpace AppResult
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{localPath} -- The local path of the file (including file name)

\item {} 
\textbf{fileName} -- The file name to use on the server

\item {} 
\textbf{directory} -- The remote directory to upload the file to on the server

\item {} 
\textbf{contentType} -- The content-type of the file

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subsection{Sample}
\label{Available modules:sample}\index{Sample (class in BaseSpacePy.model.Sample)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Sample.Sample}\pysigline{\strong{class }\code{BaseSpacePy.model.Sample.}\bfcode{Sample}}
A BaseSpace Sample object.
\index{getAccessStr() (BaseSpacePy.model.Sample.Sample method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Sample.Sample.getAccessStr}\pysiglinewithargsret{\bfcode{getAccessStr}}{\emph{scope='write'}}{}
Returns the scope-string to used for requesting BaseSpace access to the sample.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{scope} -- The scope type that is request (eg. write, read).

\end{description}\end{quote}

\end{fulllineitems}

\index{getFiles() (BaseSpacePy.model.Sample.Sample method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Sample.Sample.getFiles}\pysiglinewithargsret{\bfcode{getFiles}}{\emph{api}, \emph{queryPars=None}}{}
Returns a list of File objects
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- A BaseSpaceAPI instance

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{getReferencedAppResults() (BaseSpacePy.model.Sample.Sample method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Sample.Sample.getReferencedAppResults}\pysiglinewithargsret{\bfcode{getReferencedAppResults}}{\emph{api}}{}
Return the AppResults referenced by this sample. 
Note the returned AppResult objects do not have their ``References'' field set, 
to get a fully populated AppResult object you must use getAppResultById in BaseSpaceAPI.
If other reference types than AppResults are present (they shouldn't be), they are ignored.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{api} -- A BaseSpaceAPI instance

\item[{Returns}] \leavevmode
A list of AppResult that are referenced by this sample (with their References field not populated)

\end{description}\end{quote}

\end{fulllineitems}

\index{isInit() (BaseSpacePy.model.Sample.Sample method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Sample.Sample.isInit}\pysiglinewithargsret{\bfcode{isInit}}{}{}
Tests if the Sample instance has been initialized.
\begin{quote}\begin{description}
\item[{Raises ModelNotInitializedException}] \leavevmode
if the Id variable is not set

\item[{Returns}] \leavevmode
True on success

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subsection{File}
\label{Available modules:file}\index{File (class in BaseSpacePy.model.File)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File}\pysigline{\strong{class }\code{BaseSpacePy.model.File.}\bfcode{File}}
Represents a BaseSpace file object
\index{downloadFile() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.downloadFile}\pysiglinewithargsret{\bfcode{downloadFile}}{\emph{api}, \emph{localDir}, \emph{byteRange=None}, \emph{createBsDir=False}}{}
Download the file object to the specified localDir or a byte range of the file, by specifying the 
start and stop byte in the range.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- A BaseSpaceAPI with read access on the scope including the file object.

\item {} 
\textbf{localDir} -- The local directory to place the file in.

\item {} 
\textbf{byteRange} -- (optional) Specify the start and stop byte of the file chunk that needs retrieved (as a 2-element list).

\item {} 
\textbf{createBsDir} -- (optional) create BaseSpace File's directory inside localDir (default: False)

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{filterVariant() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.filterVariant}\pysiglinewithargsret{\bfcode{filterVariant}}{\emph{api}, \emph{Chrom}, \emph{StartPos}, \emph{EndPos}, \emph{Format='json'}, \emph{queryPars=None}}{}
List the variants in a set of variants. Note the maximum returned records is 1000.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{Chrom} -- Chromosome name

\item {} 
\textbf{StartPos} -- The start position of region of interest as a string

\item {} 
\textbf{EndPos} -- The end position of region of interest as a string

\item {} 
\textbf{Format} -- (optional) Format for results, possible values: `vcf' (not implemented yet), `json'(default, which actually returns an object)

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\item[{Returns}] \leavevmode
a list of Variant objects, when Format is json; a string, when Format is vcf

\end{description}\end{quote}

\end{fulllineitems}

\index{getCoverageMeta() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.getCoverageMeta}\pysiglinewithargsret{\bfcode{getCoverageMeta}}{\emph{api}, \emph{Chrom}}{}
Returns metadata about an alignment, including max coverage and cov granularity.        
Note that HrefCoverage must be available for the provided BAM file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI.

\item {} 
\textbf{Chrom} -- Chromosome name

\end{itemize}

\item[{Returns}] \leavevmode
a CoverageMetaData object

\end{description}\end{quote}

\end{fulllineitems}

\index{getFileS3metadata() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.getFileS3metadata}\pysiglinewithargsret{\bfcode{getFileS3metadata}}{\emph{api}}{}
Returns the S3 url and etag (md5 for small files uploaded as a single part) for a BaseSpace file
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{api} -- A BaseSpaceAPI with read access on the scope including the file object.

\item[{Returns}] \leavevmode
Dict with s3 url (`url' key) and etag (`etag' key)

\end{description}\end{quote}

\end{fulllineitems}

\index{getFileUrl() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.getFileUrl}\pysiglinewithargsret{\bfcode{getFileUrl}}{\emph{api}}{}
** Deprecated in favor of getFileS3metadata() **

Return the S3 url of the file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{api} -- A BaseSpaceAPI with read access on the scope including the file object.

\end{description}\end{quote}

\end{fulllineitems}

\index{getIntervalCoverage() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.getIntervalCoverage}\pysiglinewithargsret{\bfcode{getIntervalCoverage}}{\emph{api}, \emph{Chrom}, \emph{StartPos}, \emph{EndPos}}{}
Returns mean coverage levels over a sequence interval.
Note that HrefCoverage must be available for the provided BAM file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{Chrom} -- Chromosome name as a string - for example `chr2'

\item {} 
\textbf{StartPos} -- get coverage starting at this position

\item {} 
\textbf{EndPos} -- get coverage up to and including this position; the returned EndPos may be larger than requested due to rounding up to nearest window end coordinate

\end{itemize}

\item[{Returns}] \leavevmode
A Coverage object

\end{description}\end{quote}

\end{fulllineitems}

\index{getVariantMeta() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.getVariantMeta}\pysiglinewithargsret{\bfcode{getVariantMeta}}{\emph{api}, \emph{Format='json'}}{}
Returns the header information of a VCF file.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{Format} -- (optional) The return-value format, set to `vcf' to VCF format (string) or `json' (default, which acutally returns on object)

\end{itemize}

\item[{Returns}] \leavevmode
A VariantHeader object

\end{description}\end{quote}

\end{fulllineitems}

\index{isInit() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.isInit}\pysiglinewithargsret{\bfcode{isInit}}{}{}
Tests if the File instance has been initialized.
\begin{quote}\begin{description}
\item[{Raises ModelNotInitializedException}] \leavevmode
if the Id variable is not set

\item[{Returns}] \leavevmode
True on success

\end{description}\end{quote}

\end{fulllineitems}

\index{isValidFileOption() (BaseSpacePy.model.File.File method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.File.File.isValidFileOption}\pysiglinewithargsret{\bfcode{isValidFileOption}}{\emph{filetype}}{}~\begin{description}
\item[{** Deprecated - HrefCoverage should be present for all BAM files in BaseSpace.}] \leavevmode
However, the attribute may be missing when there has been an error
when BaseSpace internally generates coverage data from the BAM file. 
This is the same situation for HrefVariants on all VCF files. **

\end{description}

Is called to test if the File instance matches the filetype parameter
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{filetype} -- The filetype for coverage or variant requests (eg., `bam', `vcf')

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subsection{Run}
\label{Available modules:run}\index{Run (class in BaseSpacePy.model.Run)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Run.Run}\pysigline{\strong{class }\code{BaseSpacePy.model.Run.}\bfcode{Run}}
A BaseSpace Run object
\index{getAccessStr() (BaseSpacePy.model.Run.Run method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Run.Run.getAccessStr}\pysiglinewithargsret{\bfcode{getAccessStr}}{\emph{scope='write'}}{}
Returns the scope-string to used for requesting BaseSpace access to the Run.
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode
\textbf{scope} -- The scope type that is request (eg. write, read).

\end{description}\end{quote}

\end{fulllineitems}

\index{getFiles() (BaseSpacePy.model.Run.Run method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Run.Run.getFiles}\pysiglinewithargsret{\bfcode{getFiles}}{\emph{api}, \emph{queryPars=None}}{}
Returns a list of File objects associated with the Run
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{getSamples() (BaseSpacePy.model.Run.Run method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Run.Run.getSamples}\pysiglinewithargsret{\bfcode{getSamples}}{\emph{api}, \emph{queryPars=None}}{}
Returns a list of Sample objects associated with the Run
\begin{quote}\begin{description}
\item[{Parameters}] \leavevmode\begin{itemize}
\item {} 
\textbf{api} -- An instance of BaseSpaceAPI

\item {} 
\textbf{queryPars} -- An (optional) object of type QueryParameters for custom sorting and filtering

\end{itemize}

\end{description}\end{quote}

\end{fulllineitems}

\index{isInit() (BaseSpacePy.model.Run.Run method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.Run.Run.isInit}\pysiglinewithargsret{\bfcode{isInit}}{}{}
Tests if the Run instance has been initialized.
\begin{quote}\begin{description}
\item[{Raises ModelNotInitializedException}] \leavevmode
if the Id variable is not set

\item[{Returns}] \leavevmode
True on success

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\subsection{QueryParameters}
\label{Available modules:queryparameters}\index{QueryParameters (class in BaseSpacePy.model.QueryParameters)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.QueryParameters.QueryParameters}\pysiglinewithargsret{\strong{class }\code{BaseSpacePy.model.QueryParameters.}\bfcode{QueryParameters}}{\emph{pars=None}, \emph{required=None}}{}
The QueryParameters class can be passed as an optional argument
for sorting/filtering of list-responses (such as lists of samples, AppResults, variants, etc.)
\index{validate() (BaseSpacePy.model.QueryParameters.QueryParameters method)}

\begin{fulllineitems}
\phantomsection\label{Available modules:BaseSpacePy.model.QueryParameters.QueryParameters.validate}\pysiglinewithargsret{\bfcode{validate}}{}{}
Validates that query parameter keys and values are properly formed:
required keys are present, and keys and values are within the set of 
known acceptable keys/values.
\begin{quote}\begin{description}
\item[{Raises}] \leavevmode\begin{itemize}
\item {} 
\textbf{UndefinedParameterException} -- when a required parameter is not present

\item {} 
\textbf{UnknownParameterException} -- when a parameter name is not present in the list of acceptable parameters names

\item {} 
\textbf{IllegalParameterException} -- when a parameter value (with a valid name) is not present in the list of acceptable parameters values

\end{itemize}

\item[{Returns}] \leavevmode
None

\end{description}\end{quote}

\end{fulllineitems}


\end{fulllineitems}



\chapter{Getting Started}
\label{Getting Started:getting-started}\label{Getting Started::doc}

\section{Introduction}
\label{Getting Started:introduction}
\code{BaseSpacePy} is a Python based SDK to be used in the development of Apps and scripts for working with Illumina's BaseSpace cloud-computing solution for next-gen sequencing data analysis.
The primary purpose of the SDK is to provide an easy-to-use Python environment enabling developers to authenticate a user, retrieve data, and upload data/results from their own analysis to BaseSpace.

If you haven't already done so, you may wish to familiarize yourself with the BaseSpace developer documentation (\href{https://developer.basespace.illumina.com/}{https://developer.basespace.illumina.com/}) prior to working through the example scripts below.

\begin{notice}{note}{Note:}
It will be necessary to have created a BaseSpace account with a new App and have the \code{client\_key} and \code{client\_secret} codes for the App available to run a number of the following examples.
\end{notice}


\subsection{Availability}
\label{Getting Started:availability}
Version 0.1 of \code{BaseSpacePy} can be checked out here:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{git} \PYG{n}{clone} \PYG{n}{git}\PYG{n+nd}{@github.com}\PYG{p}{:}\PYG{n}{basespace}\PYG{o}{/}\PYG{n}{basespace}\PYG{o}{-}\PYG{n}{python}\PYG{o}{-}\PYG{n}{sdk}\PYG{o}{.}\PYG{n}{git}
\end{Verbatim}


\subsection{Setup}
\label{Getting Started:setup}
\emph{Requirements:} Python 2.6 with the packages `urllib2', `pycurl', `multiprocessing' and `shutil' available.

The multi-part file upload will currently only run on a unix setup.

To install `BaseSpacePy' run the `setup.py' script in the \code{src} directory (for a global install you will need to run this command with root privileges):

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{cd} \PYG{n}{basespace}\PYG{o}{-}\PYG{n}{python}\PYG{o}{-}\PYG{n}{sdk}\PYG{o}{/}\PYG{n}{src}
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install}
\end{Verbatim}

If you do not have root access, you may use the \code{-{-}prefix} option to specify the install directory (make sure this directory is in you PYTHONPATH):

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{n}{install} \PYG{o}{-}\PYG{o}{-}\PYG{n}{prefix}\PYG{o}{=}\PYG{o}{/}\PYG{n}{folder}\PYG{o}{/}\PYG{o+ow}{in}\PYG{o}{/}\PYG{n}{my}\PYG{o}{/}\PYG{n}{pythonpath}
\end{Verbatim}

For more install options type:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{python} \PYG{n}{setup}\PYG{o}{.}\PYG{n}{py} \PYG{o}{-}\PYG{o}{-}\PYG{n}{help}
\end{Verbatim}

Altenatively you may include the src directory in your PYTHONPATH by doing the following export:

\begin{Verbatim}[commandchars=\\\{\}]
export PYTHONPATH=\$PYTHONPATH:/my/path/basespace-python-sdk/src
\end{Verbatim}

or add it to the PYTHONPATH at the top of your Python scripts using BaseSpacePy:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{sys}
\PYG{n}{sys}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{/my/path/BaseSpacePy\PYGZus{}vx.x/src}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k+kn}{import} \PYG{n+nn}{BaseSpacePy}
\end{Verbatim}

To test that everything is working as expected, launch a Python prompt and try importing `BaseSpacePy':

\begin{Verbatim}[commandchars=\\\{\}]
mkallberg@ubuntu:\textasciitilde{}/\$ python
\textgreater{}\textgreater{}\textgreater{} import BaseSpacePy
\end{Verbatim}


\section{Application triggering}
\label{Getting Started:application-triggering}
This section demonstrates how to retrieve the \code{AppLaunch} object produced when a user triggers a BaseSpace App.
Further, we cover how to automatically generate the scope strings to request access to the data object (be it a project, a sample, or an analysis)
that the App was triggered to analyze.

The initial http request to our App from BaseSpace is identified by an \code{ApplicationActionId}, using this piece of information
we are able to obtain information about the user who launched the App and the data that is sought analyzed by the App.
First, we instantiate a BaseSpaceAuth object using the \code{client\_key} and \code{client\_secret} codes provided on the BaseSpace developers website when registering our App:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{BaseSpacePy.api.BaseSpaceAuth} \PYG{k+kn}{import} \PYG{n}{BaseSpaceAuth}

\PYG{c}{\PYGZsh{} initialize an authentication object using the key and secret from your app}
\PYG{c}{\PYGZsh{} Fill in with your own values}
\PYG{n}{client\PYGZus{}key}                 \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{key}\PYG{o}{\textgreater{}}
\PYG{n}{client\PYGZus{}secret}              \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{secret}\PYG{o}{\textgreater{}}
\PYG{n}{ApplicationActionId}        \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{action} \PYG{n+nb}{id}\PYG{o}{\textgreater{}}
\PYG{n}{BaseSpaceUrl}               \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{https://api.cloud-endor.illumina.com/}\PYG{l+s}{'}
\PYG{n}{version}                    \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{v1pre2/}\PYG{l+s}{'}

\PYG{c}{\PYGZsh{} First we will initialize a BaseSpace authentication object}
\PYG{n}{BSauth} \PYG{o}{=} \PYG{n}{BaseSpaceAuth}\PYG{p}{(}\PYG{n}{client\PYGZus{}key}\PYG{p}{,}\PYG{n}{client\PYGZus{}secret}\PYG{p}{,}\PYG{n}{BaseSpaceUrl}\PYG{p}{,}\PYG{n}{version}\PYG{p}{)}

\PYG{c}{\PYGZsh{} By supplying the application trigger id we can get out an AppLaunch object}
\PYG{n}{triggerObj} \PYG{o}{=} \PYG{n}{BSauth}\PYG{o}{.}\PYG{n}{getAppTrigger}\PYG{p}{(}\PYG{n}{ApplicationActionId}\PYG{p}{)}
\PYG{k}{print} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{triggerObj}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{https}\PYG{p}{:}\PYG{o}{/}\PYG{o}{/}\PYG{n}{api}\PYG{o}{.}\PYG{n}{cloud}\PYG{o}{-}\PYG{n}{endor}\PYG{o}{.}\PYG{n}{illumina}\PYG{o}{.}\PYG{n}{com}\PYG{o}{/}\PYG{n}{v1pre2}\PYG{o}{/}\PYG{n}{applicationactions}\PYG{o}{/}\PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{action} \PYG{n+nb}{id}\PYG{o}{\textgreater{}}
\end{Verbatim}

We can get the type of object the app was triggered on from the \code{getLaunchType}-method in the \code{BaseSpaceAuth} instance:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} The trigger type is a list with two items, the first a string taking the one of the values ('Projects','Samples','Analyses')}
\PYG{c}{\PYGZsh{} and the second a list of the objects of that type}
\PYG{n}{triggerType} \PYG{o}{=} \PYG{n}{triggerObj}\PYG{o}{.}\PYG{n}{getLaunchType}\PYG{p}{(}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Type of data the app was triggered on}\PYG{l+s}{"}
\PYG{k}{print} \PYG{n}{triggerType}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{We can get a handle for the user who triggered the app}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{triggerObj}\PYG{o}{.}\PYG{n}{User}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{Type} \PYG{n}{of} \PYG{n}{data} \PYG{n}{the} \PYG{n}{app} \PYG{n}{was} \PYG{n}{triggered} \PYG{n}{on}
\PYG{p}{[}\PYG{l+s}{'}\PYG{l+s}{Projects}\PYG{l+s}{'}\PYG{p}{,} \PYG{p}{[}\PYG{n}{YourProject}\PYG{p}{]}\PYG{p}{]}

\PYG{n}{We} \PYG{n}{can} \PYG{n}{get} \PYG{n}{a} \PYG{n}{handle} \PYG{k}{for} \PYG{n}{the} \PYG{n}{user} \PYG{n}{who} \PYG{n}{triggered} \PYG{n}{the} \PYG{n}{app}
\PYG{l+m+mi}{152152}\PYG{p}{:} \PYG{n}{Morten} \PYG{n}{Kallberg}
\end{Verbatim}

To start working, we will want to expand our permission scope for the trigger object so we can read and write data. The details of this process is the subject of the next section.
We end this section by demonstrating how one can easily obtain the so-called ``scope string,'' used when requesting further access, from the trigger object:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{triggerObj} \PYG{o}{=} \PYG{n}{triggerType}\PYG{p}{[}\PYG{l+m+mi}{1}\PYG{p}{]}\PYG{p}{[}\PYG{o}{-}\PYG{l+m+mi}{1}\PYG{p}{]}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The scope string for requesting write access to the trigger object is:}\PYG{l+s}{"}
\PYG{k}{print} \PYG{n}{triggerObj}\PYG{o}{.}\PYG{n}{getAccessStr}\PYG{p}{(}\PYG{n}{scope}\PYG{o}{=}\PYG{l+s}{'}\PYG{l+s}{write}\PYG{l+s}{'}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{The} \PYG{n}{scope} \PYG{n}{string} \PYG{k}{for} \PYG{n}{requesting} \PYG{n}{write} \PYG{n}{access} \PYG{n}{to} \PYG{n}{the} \PYG{n}{trigger} \PYG{n+nb}{object} \PYG{o+ow}{is}\PYG{p}{:}
\PYG{n}{write} \PYG{n}{project} \PYG{l+m+mi}{89}
\end{Verbatim}


\section{Requesting an access-token for data browsing}
\label{Getting Started:requesting-an-access-token-for-data-browsing}
Here we demonstrate the basic BaseSpace authentication process. The work-flow outlined here is
\begin{enumerate}
\item {} 
Request of access to a specific data-scope

\item {} 
User approval of access request

\item {} 
Browsing data

\end{enumerate}

\begin{notice}{note}{Note:}
It will be useful if you are logged in to the BaseSpace web-site before launching this example to make the access grant procedure faster.
\end{notice}

Again we will start out by initializing a \code{BaseSpaceAuth} object:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{BaseSpacePy.api.BaseSpaceAuth} \PYG{k+kn}{import} \PYG{n}{BaseSpaceAuth}
\PYG{k+kn}{import} \PYG{n+nn}{time}

\PYG{c}{\PYGZsh{} initialize an authentication object using the key and secret from your app}
\PYG{n}{client\PYGZus{}key}                 \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{key}\PYG{o}{\textgreater{}}
\PYG{n}{client\PYGZus{}secret}              \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{secret}\PYG{o}{\textgreater{}}
\PYG{n}{BaseSpaceUrl}               \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{https://api.cloud-endor.illumina.com/}\PYG{l+s}{'}
\PYG{n}{version}                    \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{v1pre2/}\PYG{l+s}{'}
\PYG{n}{BSauth} \PYG{o}{=} \PYG{n}{BaseSpaceAuth}\PYG{p}{(}\PYG{n}{client\PYGZus{}key}\PYG{p}{,}\PYG{n}{client\PYGZus{}secret}\PYG{p}{,}\PYG{n}{BaseSpaceUrl}\PYG{p}{,}\PYG{n}{version}\PYG{p}{)}
\end{Verbatim}

First get verification code and uri for scope `browse global'

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{deviceInfo} \PYG{o}{=} \PYG{n}{BSauth}\PYG{o}{.}\PYG{n}{getVerificationCode}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{browse global}\PYG{l+s}{'}\PYG{p}{)}
\end{Verbatim}

At this point the user must visit the verification uri to grant us access

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{}\PYGZsh{} PAUSE HERE}
\PYG{c}{\PYGZsh{} Have the user visit the verification uri to grant us access}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{Please visit the uri within 30 seconds and grant access}\PYG{l+s}{"}
\PYG{k}{print} \PYG{n}{deviceInfo}\PYG{p}{[}\PYG{l+s}{'}\PYG{l+s}{verification\PYGZus{}with\PYGZus{}code\PYGZus{}uri}\PYG{l+s}{'}\PYG{p}{]}
\PYG{n}{time}\PYG{o}{.}\PYG{n}{sleep}\PYG{p}{(}\PYG{l+m+mi}{30}\PYG{p}{)}
\PYG{c}{\PYGZsh{}\PYGZsh{} PAUSE HERE}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
Output[]:

Please visit the uri within 10 seconds and grant access
https://cloud-endor.illumina.com/oauth/device?code=\textless{}my device code\textgreater{}
\end{Verbatim}

There are two options for obtaining the access-token and instantiating a \code{BaseSpaceAPI} object:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} Get the access-token directly and instantiate an api yourself}
\PYG{c}{\PYGZsh{}token = BSauth.getAccessToken(deviceInfo['device\PYGZus{}code'])}
\PYG{c}{\PYGZsh{}print "My token " + str(token)}

\PYG{c}{\PYGZsh{} Alternatively we can generate an access-token and request a BaseSpaceApi instance}
\PYG{c}{\PYGZsh{} with the newly generated token in one step}
\PYG{n}{myAPI} \PYG{o}{=} \PYG{n}{BSauth}\PYG{o}{.}\PYG{n}{getBaseSpaceApi}\PYG{p}{(}\PYG{n}{deviceInfo}\PYG{p}{[}\PYG{l+s}{'}\PYG{l+s}{device\PYGZus{}code}\PYG{l+s}{'}\PYG{p}{]}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{myAPI}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{BaseSpaceAPI} \PYG{n}{instance} \PYG{o}{-} \PYG{n}{using} \PYG{n}{token}\PYG{o}{=}\PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{access} \PYG{n}{token}\PYG{o}{\textgreater{}}
\end{Verbatim}

At this point we can start using the \code{BaseSpaceAPI} instance to browse the available data for the current user, the details of this process is the subject of the next section. Here we will end with showing how the API object can be used to list all BaseSpace genome instances:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} We will get all available genomes with our new api!}
\PYG{n}{allGenomes}  \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getAvailableGenomes}\PYG{p}{(}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Genomes }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{allGenomes}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{Genomes}
\PYG{p}{[}\PYG{n}{Arabidopsis} \PYG{n}{thaliana}\PYG{p}{,} \PYG{n}{Bos} \PYG{n}{Taurus}\PYG{p}{,} \PYG{n}{Escherichia} \PYG{n}{coli}\PYG{p}{,} \PYG{n}{Homo} \PYG{n}{sapiens}\PYG{p}{,} \PYG{n}{Mus} \PYG{n}{musculus}\PYG{p}{,} \PYG{n}{Phix}\PYG{p}{,}\PYGZbs{}
 \PYG{n}{Rhodobacter} \PYG{n}{sphaeroides}\PYG{p}{,} \PYG{n}{Rattus} \PYG{n}{norvegicus}\PYG{p}{,} \PYG{n}{Saccharomyces} \PYG{n}{cerevisiae}\PYG{p}{,} \PYG{n}{Staphylococcus} \PYG{n}{aureus}\PYG{p}{,} \PYG{n}{Bacillus} \PYG{n}{Cereus}\PYG{p}{]}
\end{Verbatim}


\section{Browsing data with \texttt{global browse}-access}
\label{Getting Started:browsing-data-with-global-browse-access}
This section demonstrates basic browsing of BaseSpace objects once an access-token for global browsing has been obtained. We will see how
objects can be retrieved using either the \code{BaseSpaceAPI} class or by use of method calls on related object instances (for example once
a \code{user} instance we can use it to retrieve all project belonging to that user).

First we will initialize a \code{BaseSpaceAPI} using our access-token for \code{global browse}:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{from} \PYG{n+nn}{BaseSpacePy.api.BaseSpaceAPI} \PYG{k+kn}{import} \PYG{n}{BaseSpaceAPI}

\PYG{c}{\PYGZsh{} REST server information and user access token}
\PYG{n}{server}          \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{https://api.cloud-endor.illumina.com/}\PYG{l+s}{'}
\PYG{n}{version}         \PYG{o}{=} \PYG{l+s}{'}\PYG{l+s}{v1pre2}\PYG{l+s}{'}
\PYG{n}{access\PYGZus{}token}    \PYG{o}{=} \PYG{o}{\textless{}}\PYG{n}{my} \PYG{n}{access} \PYG{n}{token}\PYG{o}{\textgreater{}}

\PYG{c}{\PYGZsh{} First, create a client for making calls for this user session}
\PYG{n}{myAPI}   \PYG{o}{=} \PYG{n}{BaseSpaceAPI}\PYG{p}{(}\PYG{n}{AccessToken}\PYG{o}{=}\PYG{n}{access\PYGZus{}token}\PYG{p}{,}\PYG{n}{apiServer}\PYG{o}{=} \PYG{n}{server} \PYG{o}{+} \PYG{n}{version}\PYG{p}{)}
\end{Verbatim}

First we will try to retrieve a genome object:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} Now grab the genome with id=4}
\PYG{n}{myGenome}    \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getGenomeById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{4}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The Genome is }\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{myGenome}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{We can get more information from the genome object}\PYG{l+s}{"}
\PYG{k}{print} \PYG{l+s}{'}\PYG{l+s}{Id: }\PYG{l+s}{'} \PYG{o}{+} \PYG{n}{myGenome}\PYG{o}{.}\PYG{n}{Id}
\PYG{k}{print} \PYG{l+s}{'}\PYG{l+s}{Href: }\PYG{l+s}{'} \PYG{o}{+} \PYG{n}{myGenome}\PYG{o}{.}\PYG{n}{Href}
\PYG{k}{print} \PYG{l+s}{'}\PYG{l+s}{DisplayName: }\PYG{l+s}{'} \PYG{o}{+} \PYG{n}{myGenome}\PYG{o}{.}\PYG{n}{DisplayName}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
Output[]:

The Genome is Homo sapiens
We can get more information from the genome object
Id: 4
Href: v1pre2/genomes/4
DisplayName: Homo Sapiens - UCSC (hg19)
\end{Verbatim}

Using a comparable method we can get a list of all available genomes:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} Get a list of all genomes}
\PYG{n}{allGenomes}  \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getAvailableGenomes}\PYG{p}{(}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Genomes }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{allGenomes}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{Genomes}
\PYG{p}{[}\PYG{n}{Arabidopsis} \PYG{n}{thaliana}\PYG{p}{,} \PYG{n}{Bos} \PYG{n}{Taurus}\PYG{p}{,} \PYG{n}{Escherichia} \PYG{n}{coli}\PYG{p}{,} \PYG{n}{Homo} \PYG{n}{sapiens}\PYG{p}{,} \PYG{n}{Mus} \PYG{n}{musculus}\PYG{p}{,} \PYG{n}{Phix}\PYG{p}{,}\PYGZbs{}
 \PYG{n}{Rhodobacter} \PYG{n}{sphaeroides}\PYG{p}{,} \PYG{n}{Rattus} \PYG{n}{norvegicus}\PYG{p}{,} \PYG{n}{Saccharomyces} \PYG{n}{cerevisiae}\PYG{p}{,} \PYG{n}{Staphylococcus} \PYG{n}{aureus}\PYG{p}{,} \PYG{n}{Bacillus} \PYG{n}{Cereus}\PYG{p}{]}
\end{Verbatim}

Now, let us retrieve the \code{User} objects for the current user, and list all projects for this user:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} Take a look at the current user}
\PYG{n}{user}        \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getUserById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{current}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The current user is }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{user}\PYG{p}{)}

\PYG{c}{\PYGZsh{} Now list the projects for this user}
\PYG{n}{myProjects}   \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getProjectByUser}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{current}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The projects for this user are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{myProjects}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{The} \PYG{n}{current} \PYG{n}{user} \PYG{o+ow}{is}
\PYG{l+m+mi}{152152}\PYG{p}{:} \PYG{n}{Morten} \PYG{n}{Kallberg}

\PYG{n}{The} \PYG{n}{projects} \PYG{k}{for} \PYG{n}{this} \PYG{n}{user} \PYG{n}{are}
\PYG{p}{[}\PYG{n}{HiSeq} \PYG{l+m+mi}{2500}\PYG{p}{,} \PYG{n}{Bolt}\PYG{p}{,} \PYG{n}{YourProject}\PYG{p}{,} \PYG{l+m+mi}{2}\PYG{n}{X151} \PYG{n}{Rhodobacter} \PYG{n}{Resequencing}\PYG{p}{,} \PYG{n}{EColi} \PYG{n}{resequencing}\PYG{p}{]}
\end{Verbatim}

We can also achieve this by making a call using the \code{user} instance. Notice that these calls take an instance of \code{BaseSpaceAPI} with apporpriate
priviliges to complete the transaction as parameter, this true for all retrieval method calls made on data objects:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{myProjects2} \PYG{o}{=} \PYG{n}{user}\PYG{o}{.}\PYG{n}{getProjects}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Projects retrieved from the user instance }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{myProjects2}\PYG{p}{)}

\PYG{c}{\PYGZsh{} List the runs available for the current user}
\PYG{n}{runs} \PYG{o}{=} \PYG{n}{user}\PYG{o}{.}\PYG{n}{getRuns}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The runs for this user are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{runs}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
Output[]:

Projects retrieved from the user instance
[HiSeq 2500, Bolt, YourProject, 2X151 Rhodobacter Resequencing, EColi resequencing]

The runs for this user are
[2X151 Rhodobacter Resequencing, 2x26 Validation Human 4-Plex, EColi resequencing]
\end{Verbatim}

In the same manner we can get a list of accessible user runs:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{runs} \PYG{o}{=} \PYG{n}{user}\PYG{o}{.}\PYG{n}{getRuns}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Runs retrieved from user instance }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{runs}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
Output[]:

Runs retrieved from user instance
[2X151 Rhodobacter Resequencing, 2x26 Validation Human 4-Plex, EColi resequencing]
\end{Verbatim}


\section{Accessing file-trees and querying BAM/VCF files}
\label{Getting Started:accessing-file-trees-and-querying-bam-vcf-files}
In this section we demonstrate how to access samples and analysis from a projects and how to work with the available file data for such instances.
In addition, we take a look at some of the special queuring methods associated with BAM- and VCF-files.

Again, start out by initializing a \code{BaseSpacePy} instance and retrieving all projects belonging to the current user:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} First, create a client for making calls for this user session}
\PYG{n}{myAPI}   \PYG{o}{=} \PYG{n}{BaseSpaceAPI}\PYG{p}{(}\PYG{n}{AccessToken}\PYG{o}{=}\PYG{n}{access\PYGZus{}token}\PYG{p}{,}\PYG{n}{apiServer}\PYG{o}{=} \PYG{n}{server} \PYG{o}{+} \PYG{n}{version}\PYG{p}{)}
\PYG{n}{user}        \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getUserById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{current}\PYG{l+s}{'}\PYG{p}{)}
\PYG{n}{myProjects}   \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getProjectByUser}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{current}\PYG{l+s}{'}\PYG{p}{)}
\end{Verbatim}

Now we can list all the analyses and samples for these projects

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k}{for} \PYG{n}{singleProject} \PYG{o+ow}{in} \PYG{n}{myProjects}\PYG{p}{:}
    \PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{\PYGZsh{} }\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{singleProject}\PYG{p}{)}
    \PYG{n}{analyses} \PYG{o}{=} \PYG{n}{singleProject}\PYG{o}{.}\PYG{n}{getAnalyses}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
    \PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{    The analysis for project }\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{singleProject}\PYG{p}{)} \PYG{o}{+} \PYG{l+s}{"}\PYG{l+s}{ are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+se}{\PYGZbs{}t}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{analyses}\PYG{p}{)}
    \PYG{n}{samples} \PYG{o}{=} \PYG{n}{singleProject}\PYG{o}{.}\PYG{n}{getSamples}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
    \PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{    The samples for project }\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{singleProject}\PYG{p}{)} \PYG{o}{+} \PYG{l+s}{"}\PYG{l+s}{ are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s+se}{\PYGZbs{}t}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{samples}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{c}{\PYGZsh{} HiSeq 2500}
    \PYG{n}{The} \PYG{n}{analysis} \PYG{k}{for} \PYG{n}{project} \PYG{n}{HiSeq} \PYG{l+m+mi}{2500} \PYG{n}{are}
        \PYG{p}{[}\PYG{n}{Resequencing} \PYG{o}{-} \PYG{n}{Completed}\PYG{p}{]}
    \PYG{n}{The} \PYG{n}{samples} \PYG{k}{for} \PYG{n}{project} \PYG{n}{HiSeq} \PYG{l+m+mi}{2500} \PYG{n}{are}
        \PYG{p}{[}\PYG{n}{NA18507}\PYG{p}{]}
\PYG{c}{\PYGZsh{} Bolt}
    \PYG{n}{The} \PYG{n}{analysis} \PYG{k}{for} \PYG{n}{project} \PYG{n}{Bolt} \PYG{n}{are}
        \PYG{p}{[}\PYG{n}{Amplicon} \PYG{o}{-} \PYG{n}{Completed}\PYG{p}{,} \PYG{n}{Amplicon} \PYG{o}{-} \PYG{n}{Completed}\PYG{p}{,} \PYG{n}{Amplicon} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}
    \PYG{n}{The} \PYG{n}{samples} \PYG{k}{for} \PYG{n}{project} \PYG{n}{Bolt} \PYG{n}{are}
        \PYG{p}{[}\PYG{n}{sample\PYGZus{}1}\PYG{p}{,} \PYG{n}{sample\PYGZus{}2}\PYG{p}{,} \PYG{n}{sample\PYGZus{}3}\PYG{p}{,} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}

\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}
\end{Verbatim}

We'll take a further look at the files belonging to the sample from the last project in the loop above:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k}{for} \PYG{n}{s} \PYG{o+ow}{in} \PYG{n}{samples}\PYG{p}{:}
    \PYG{k}{print} \PYG{l+s}{"}\PYG{l+s}{Sample }\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{s}\PYG{p}{)}
    \PYG{n}{ff} \PYG{o}{=} \PYG{n}{s}\PYG{o}{.}\PYG{n}{getFiles}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
    \PYG{k}{print} \PYG{n}{ff}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{Sample} \PYG{n}{Ecoli}
\PYG{p}{[}\PYG{n}{s\PYGZus{}G1\PYGZus{}L001\PYGZus{}R1\PYGZus{}001}\PYG{o}{.}\PYG{n}{fastq}\PYG{o}{.}\PYG{l+m+mf}{1.}\PYG{n}{gz}\PYG{p}{,} \PYG{n}{s\PYGZus{}G1\PYGZus{}L001\PYGZus{}R1\PYGZus{}002}\PYG{o}{.}\PYG{n}{fastq}\PYG{o}{.}\PYG{l+m+mf}{1.}\PYG{n}{gz}\PYG{p}{,} \PYG{n}{s\PYGZus{}G1\PYGZus{}L001\PYGZus{}R2\PYGZus{}001}\PYG{o}{.}\PYG{n}{fastq}\PYG{o}{.}\PYG{l+m+mf}{1.}\PYG{n}{gz}\PYG{p}{,} \PYG{n}{s\PYGZus{}G1\PYGZus{}L001\PYGZus{}R2\PYGZus{}002}\PYG{o}{.}\PYG{n}{fastq}\PYG{o}{.}\PYG{l+m+mf}{1.}\PYG{n}{gz}\PYG{p}{]}
\end{Verbatim}

Now, have a look at some of the methods calls specific to \code{Bam} and \code{VCF} files. First, we will get a \code{Bam}-file and then retrieve the coverage information available for chromosome 2 between positions 1 and 20000:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} Now do some work with files}
\PYG{c}{\PYGZsh{} we'll grab a BAM by id and get the coverage for an interval + accompanying meta-data}
\PYG{n}{myBam} \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getFileById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{2150156}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{myBam}
\PYG{n}{cov}     \PYG{o}{=} \PYG{n}{myBam}\PYG{o}{.}\PYG{n}{getIntervalCoverage}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{chr2}\PYG{l+s}{'}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{1}\PYG{l+s}{'}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{20000}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{cov}
\PYG{n}{covMeta} \PYG{o}{=} \PYG{n}{myBam}\PYG{o}{.}\PYG{n}{getCoverageMeta}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{chr2}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{covMeta}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n+nb}{sorted}\PYG{o}{.}\PYG{n}{bam}
\PYG{n}{Chrchr2}\PYG{p}{:} \PYG{l+m+mi}{1}\PYG{o}{-}\PYG{l+m+mi}{20096}\PYG{p}{:} \PYG{n}{BucketSize}\PYG{o}{=}\PYG{l+m+mi}{16}
\PYG{n}{CoverageMeta}\PYG{p}{:} \PYG{n+nb}{max}\PYG{o}{=}\PYG{l+m+mi}{20483} \PYG{n}{gran}\PYG{o}{=}\PYG{l+m+mi}{128}
\end{Verbatim}

For \code{VCF}-files we can filter variant calls based on chromosome and location as well:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{c}{\PYGZsh{} and a vcf file}
\PYG{n}{myVCF} \PYG{o}{=} \PYG{n}{myAPI}\PYG{o}{.}\PYG{n}{getFileById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{2150158}\PYG{l+s}{'}\PYG{p}{)}
\PYG{c}{\PYGZsh{} Get the variant meta info}
\PYG{n}{varMeta} \PYG{o}{=} \PYG{n}{myVCF}\PYG{o}{.}\PYG{n}{getVariantMeta}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{varMeta}
\PYG{n}{var}     \PYG{o}{=} \PYG{n}{myVCF}\PYG{o}{.}\PYG{n}{filterVariant}\PYG{p}{(}\PYG{n}{myAPI}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{2}\PYG{l+s}{'}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{1}\PYG{l+s}{'}\PYG{p}{,} \PYG{l+s}{'}\PYG{l+s}{11000}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{n}{var}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{VariantHeader}\PYG{p}{:} \PYG{n}{SampleCount}\PYG{o}{=}\PYG{l+m+mi}{1}
\PYG{p}{[}\PYG{n}{Variant} \PYG{o}{-} \PYG{n}{chr2}\PYG{p}{:} \PYG{l+m+mi}{10236} \PYG{n+nb}{id}\PYG{o}{=}\PYG{p}{[}\PYG{l+s}{'}\PYG{l+s}{.}\PYG{l+s}{'}\PYG{p}{]}\PYG{p}{,} \PYG{n}{Variant} \PYG{o}{-} \PYG{n}{chr2}\PYG{p}{:} \PYG{l+m+mi}{10249} \PYG{n+nb}{id}\PYG{o}{=}\PYG{p}{[}\PYG{l+s}{'}\PYG{l+s}{.}\PYG{l+s}{'}\PYG{p}{]}\PYG{p}{,} \PYG{o}{.}\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}
\end{Verbatim}


\section{Creating an analysis and uploading results}
\label{Getting Started:creating-an-analysis-and-uploading-results}
In this section we will see how to create a new analysis object, change its state
and upload result files to it as well as retrieve files from it.

First, create a client for making calls for this user session:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{myBaseSpaceAPI}   \PYG{o}{=} \PYG{n}{BaseSpaceAPI}\PYG{p}{(}\PYG{n}{AccessToken}\PYG{o}{=}\PYG{n}{access\PYGZus{}token}\PYG{p}{,}\PYG{n}{apiServer}\PYG{o}{=} \PYG{n}{server} \PYG{o}{+} \PYG{n}{version}\PYG{p}{)}
\PYG{c}{\PYGZsh{}}
\PYG{c}{\PYGZsh{}\PYGZsh{} Now we'll do some work of our own. First get a project to work on}
\PYG{c}{\PYGZsh{}\PYGZsh{} we'll need write permission, for the project we are working on}
\PYG{c}{\PYGZsh{}\PYGZsh{} meaning we will need get a new token and instantiate a new BaseSpaceAPI}
\PYG{n}{p} \PYG{o}{=} \PYG{n}{myBaseSpaceAPI}\PYG{o}{.}\PYG{n}{getProjectById}\PYG{p}{(}\PYG{l+s}{'}\PYG{l+s}{89}\PYG{l+s}{'}\PYG{p}{)}
\PYG{c}{\PYGZsh{} A short-cut for getting a scope string if we already have a project-instance:}
\PYG{k}{print} \PYG{n}{p}\PYG{o}{.}\PYG{n}{getAccessStr}\PYG{p}{(}\PYG{n}{scope}\PYG{o}{=}\PYG{l+s}{'}\PYG{l+s}{write}\PYG{l+s}{'}\PYG{p}{)}
\PYG{c}{\PYGZsh{} or simply}
\PYG{n}{p}\PYG{o}{.}\PYG{n}{getAccessStr}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{write} \PYG{n}{project} \PYG{l+m+mi}{89}
\end{Verbatim}

Assuming we now have write access for the project, we will list the current analyses for the project:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{ana} \PYG{o}{=} \PYG{n}{p}\PYG{o}{.}\PYG{n}{getAnalyses}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The current analyses are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{ana}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{The} \PYG{n}{current} \PYG{n}{analyses} \PYG{n}{are}
\PYG{p}{[}\PYG{n}{Results} \PYG{k}{for} \PYG{n}{sample} \PYG{l+m+mi}{123} \PYG{o}{-} \PYG{n}{Working}\PYG{p}{,} \PYG{n}{Results} \PYG{k}{for} \PYG{n}{sample} \PYG{l+m+mi}{124} \PYG{o}{-} \PYG{n}{Working}\PYG{o}{.}\PYG{o}{.}\PYG{o}{.}
\end{Verbatim}

To create an analysis for a project, simply give the name and description to the \code{createAnalysis}-method:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{analysis} \PYG{o}{=} \PYG{n}{p}\PYG{o}{.}\PYG{n}{createAnalysis}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{,}\PYG{l+s}{"}\PYG{l+s}{My very first analysis!!}\PYG{l+s}{"}\PYG{p}{,}\PYG{l+s}{"}\PYG{l+s}{This is my analysis}\PYG{l+s}{"}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{Some info about our new analysis}\PYG{l+s}{"}
\PYG{k}{print} \PYG{n}{analysis}
\PYG{k}{print} \PYG{n}{analysis}\PYG{o}{.}\PYG{n}{Id}
\PYG{k}{print} \PYG{n}{analysis}\PYG{o}{.}\PYG{n}{Status}
\PYG{c}{\PYGZsh{} we can change the status of out analysis and add a status-summary as follows}
\PYG{n}{analysis}\PYG{o}{.}\PYG{n}{setStatus}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{completed}\PYG{l+s}{'}\PYG{p}{,}\PYG{l+s}{"}\PYG{l+s}{We worked hard.}\PYG{l+s}{"}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{After a change of status we get}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{analysis}\PYG{p}{)}

\PYG{c}{\PYGZsh{}\PYGZsh{}\PYGZsh{} List the analyses again and see if our new object shows up}
\PYG{n}{ana} \PYG{o}{=} \PYG{n}{p}\PYG{o}{.}\PYG{n}{getAnalyses}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{The updated analyses are }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{ana}\PYG{p}{)}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
Output[]:

Some info about our new analysis
My very first analysis!! - Working
94094
Working

After a change of status we get
My very first analysis!! - Completed

The updated analyses are
[Results for sample 123 - Working, Results for sample 124 - Working, Results for sample 124 - Working, ...
\end{Verbatim}

Now we will make another analysis and try to upload some files to it:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{analysis2} \PYG{o}{=} \PYG{n}{p}\PYG{o}{.}\PYG{n}{createAnalysis}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{,}\PYG{l+s}{"}\PYG{l+s}{My second analysis}\PYG{l+s}{"}\PYG{p}{,}\PYG{l+s}{"}\PYG{l+s}{This one I will upload to}\PYG{l+s}{"}\PYG{p}{)}
\PYG{n}{analysis2}\PYG{o}{.}\PYG{n}{uploadFile}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{,} \PYG{l+s}{'}\PYG{l+s}{/my/file/dir/testFile2.txt}\PYG{l+s}{'}\PYG{p}{,} \PYG{l+s}{'}\PYG{l+s}{BaseSpaceTestFile.txt}\PYG{l+s}{'}\PYG{p}{,} \PYG{l+s}{'}\PYG{l+s}{/mydir/}\PYG{l+s}{'}\PYG{p}{,} \PYG{l+s}{'}\PYG{l+s}{text/plain}\PYG{l+s}{'}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{My analysis number 2 }\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{"} \PYG{o}{+} \PYG{n+nb}{str}\PYG{p}{(}\PYG{n}{analysis2}\PYG{p}{)}
\PYG{c}{\PYGZsh{}}
\PYG{c}{\PYGZsh{}\PYGZsh{} Check to see if our new file made it}
\PYG{n}{analysisFiles} \PYG{o}{=} \PYG{n}{analysis2}\PYG{o}{.}\PYG{n}{getFiles}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{)}
\PYG{k}{print} \PYG{l+s}{"}\PYG{l+s+se}{\PYGZbs{}n}\PYG{l+s}{These are the files in the analysis}\PYG{l+s}{"}
\PYG{k}{print} \PYG{n}{analysisFiles}
\PYG{n}{f} \PYG{o}{=} \PYG{n}{analysisFiles}\PYG{p}{[}\PYG{o}{-}\PYG{l+m+mi}{1}\PYG{p}{]}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Output}\PYG{p}{[}\PYG{p}{]}\PYG{p}{:}

\PYG{n}{My} \PYG{n}{analysis} \PYG{n}{number} \PYG{l+m+mi}{2}
\PYG{n}{My} \PYG{n}{second} \PYG{n}{analysis} \PYG{o}{-} \PYG{n}{Working}

\PYG{n}{These} \PYG{n}{are} \PYG{n}{the} \PYG{n}{files} \PYG{o+ow}{in} \PYG{n}{the} \PYG{n}{analysis}
\PYG{p}{[}\PYG{n}{BaseSpaceTestFile}\PYG{o}{.}\PYG{n}{txt}\PYG{p}{]}
\end{Verbatim}

We can even download our newly uploaded file in the following manner:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{f}\PYG{o}{.}\PYG{n}{downloadFile}\PYG{p}{(}\PYG{n}{myBaseSpaceAPI}\PYG{p}{,}\PYG{l+s}{'}\PYG{l+s}{/path/to/place/file/in/}\PYG{l+s}{'}\PYG{p}{)}
\end{Verbatim}


\chapter{Indices and tables}
\label{index:indices-and-tables}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}
